streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -406,6 +407,23 @@ final class ZSink[-R, +E, -In, +L, +Z] private (val channel: ZChannel[R, ZNothin
     ZSink.unwrapScoped(scoped)
   }
 
+  def refineOrDie[E1](
+    pf: PartialFunction[E, E1]
+  )(implicit ev1: E <:< Throwable, ev2: CanFail[E], trace: Trace): ZSink[R, E1, In, L, Z] =
+    refineOrDieWith(pf)(identity(_))
+
+  def refineOrDieWith[E1](
+    pf: PartialFunction[E, E1]
+  )(f: E => Throwable)(implicit ev: CanFail[E], trace: Trace): ZSink[R, E1, In, L, Z] =
+    new ZSink(
+      channel.catchAll(e =>
+        if (pf.isDefinedAt(e))
```

Comment:
This implementation currently requires potentially evaluating the `partialFunction` twice (once in `isDefinedAt` and a second time in `apply`). We should use another operator here like `applyOrElse` to avoid this.

Reply:
Updated here and also for ZStream